### PR TITLE
FMBench: benchmarking Llama models on AWS

### PR DESCRIPTION
**Pull request by aarora79**
# What does this PR do?

Adds a recipe (`FMBench`) for benchmarking Llama models (including Llama3) on AWS platforms (SageMaker, Bedrock). `FMBench` is an open-source Python package for benchmarking foundation models for price|performance. See [`FMBench`](https://github.com/aws-samples/foundation-model-benchmarking-tool/tree/main).

Fixes # (issue)

## Feature/Issue validation/testing

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
Logs for Test A

- [ ] Test B
Logs for Test B

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x ] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ x] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

File: recipes/benchmarks/fmbench/README.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+<sup>1</sup> ** represent values hidden on purpose, these are available when you run the tool yourself.
+
+The report also includes latency Vs prompt size charts for different concurrency levels. As expected, inference latency increases as prompt size increases but what is interesting to note is that the increase is much more at higher concurrency levels (and this behavior varies with instance types).
+
+![Effect of prompt size on inference latency for different concurrency levels](./img/latency_vs_tokens.png)
+
+### How to get started with `FMBench`
+
+The following steps provide a Quick start guide for `FMBench`. For a more detailed DIY version, please see the [`FMBench Readme`](https://github.com/aws-samples/foundation-model-benchmarking-tool?tab=readme-ov-file#the-diy-version-with-gory-details).
```
Comment (HamidShojanazeri) [inline recipes/benchmarks/fmbench/README.md:45]:
```suggestion
The following steps provide a [Quick start guide for `FMBench`](https://github.com/aws-samples/foundation-model-benchmarking-tool#quickstart). For a more detailed DIY version, please see the [`FMBench Readme`](https://github.com/aws-samples/foundation-model-benchmarking-tool?tab=readme-ov-file#the-diy-version-with-gory-details).
```

Comment (HamidShojanazeri) [reply recipes/benchmarks/fmbench/README.md:45]: Wonder DIY section is part of quickstart? or if someone does quick start would not need it?

Comment (HamidShojanazeri) [reply recipes/benchmarks/fmbench/README.md:45]: Maybe good to clarify if quick start handles the following steps ( step1?)

File: recipes/benchmarks/fmbench/README.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+1. On the `fmbench-notebook` open a Terminal and run the following commands.
+
+    ```{.bash}
+    conda create --name fmbench_python311 -y python=3.11 ipykernel
+    source activate fmbench_python311;
+    pip install -U fmbench
+    ```
+
+1. Now you are ready to `fmbench` with the following command line. We will use a sample config file placed in the S3 bucket by the CloudFormation stack for a quick first run.
```
Comment (HamidShojanazeri) [inline recipes/benchmarks/fmbench/README.md:64]: numbered list update

Comment (HamidShojanazeri) [reply recipes/benchmarks/fmbench/README.md:64]: also it has format issues.

File: recipes/benchmarks/fmbench/README.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    1. We benchmark performance for the `Llama2-7b` model on a `ml.g5.xlarge` and a `ml.g5.2xlarge` instance type, using the `huggingface-pytorch-tgi-inference` inference container. This test would take about 30 minutes to complete and cost about $0.20.
+
+    1. It uses a simple relationship of 750 words equals 1000 tokens, to get a more accurate representation of token counts use the `Llama2 tokenizer` (instructions are provided in the next section). ***It is strongly recommended that for more accurate results on token throughput you use a tokenizer specific to the model you are testing rather than the default tokenizer. See instructions provided later in this document on how to use a custom tokenizer***.
```
Comment (HamidShojanazeri) [inline recipes/benchmarks/fmbench/README.md:68]: can we clarify this part a bit, are we packing the requests at 1k tokens? how batching works here? is it packing instead of batching?

Comment (HamidShojanazeri) [reply recipes/benchmarks/fmbench/README.md:68]: > See instructions provided later in this document on how to use a custom

Shall we pls add a link to where we are providing the instructions?

File: recipes/benchmarks/fmbench/README.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    1. Open another terminal window and do a `tail -f` on the `fmbench.log` file to see all the traces being generated at runtime.
+
+        ```{.bash}
+        tail -f fmbench.log
+        ```
+
+1. The generated reports and metrics are available in the `sagemaker-fmbench-write-<replace_w_your_aws_region>-<replace_w_your_aws_account_id>` bucket. The metrics and report files are also downloaded locally and in the `results` directory (created by `FMBench`) and the benchmarking report is available as a markdown file called `report.md` in the `results` directory. You can view the rendered Markdown report in the SageMaker notebook itself or download the metrics and report files to your machine for offline analysis.
+
+## The `config.yml` file
```
Comment (HamidShojanazeri) [inline recipes/benchmarks/fmbench/README.md:84]: should we mention config file somewhere in the top, that all you need to do is set up a config file and then use our automated process to launch the job?

File: recipes/benchmarks/fmbench/README.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+- [Config file](https://github.com/aws-samples/foundation-model-benchmarking-tool/blob/main/src/fmbench/configs/config-llama2-70b-g5-p4d-trt.yml) for `Llama2-70b` on `ml.p4d.24xlarge` instance using the [Deep Java Library TensorRT container](763104351884.dkr.ecr.us-east-1.amazonaws.com/djl-inference:0.26.0-tensorrtllm0.7.1-cu122).
+- [Config file](https://github.com/aws-samples/foundation-model-benchmarking-tool/blob/main/src/fmbench/configs/config-llama2-70b-inf2-g5.yml) for `Llama2-70b` on `ml.inf2.48xlarge` instance using the [HuggingFace TGI with Optimum NeuronX container](763104351884.dkr.ecr.{region}.amazonaws.com/huggingface-pytorch-tgi-inference:1.13.1-optimum0.0.17-neuronx-py310-ubuntu22.04).
+
+## Benchmarking Llama2 on Amazon Bedrock
```
Comment (HamidShojanazeri) [inline recipes/benchmarks/fmbench/README.md:107]: Llama3 just got added, I wonder we can update this part too.

**Review by HamidShojanazeri — COMMENTED** [review]
@aarora79 thanks a lot for the PR! left some  minor comments inline.

can you also pls sign the CLA.

File: recipes/benchmarks/fmbench/README.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+`FMBench` runs inference requests against endpoints that are either deployed through `FMBench` itself (as in the case of SageMaker) or are available either as a fully-managed endpoint (as in the case of Bedrock) or as bring your own endpoint. The metrics such as inference latency, transactions per-minute, error rates and cost per transactions are captured and presented in the form of a Markdown report containing explanatory text, tables and figures. The figures and tables in the report provide insights into what might be the best serving stack (instance type, inference container and configuration parameters) for a given Llama model for a given use-case.
+
+The following figure gives an example of the price performance numbers that include inference latency, transactions per-minute and concurrency level for running the `Llama2-13b` model on different instance types available on SageMaker using prompts for Q&A task created from the [`LongBench`](https://huggingface.co/datasets/THUDM/LongBench) dataset, these prompts are between 3000 to 3840 tokens in length. **_Note that the numbers are hidden in this figure but you would be able to see them when you run `FMBench` yourself_**.
```
[the remark this answers is not in the document]
Comment (aarora79) [reply recipes/benchmarks/fmbench/README.md:13]: HF datasets are supported through the [`bring_your_own_dataset.ipynb`](https://github.com/aws-samples/foundation-model-benchmarking-tool/blob/main/src/fmbench/bring_your_own_dataset.ipynb) notebook but technically you could tweak that code and support any dataset.

File: recipes/benchmarks/fmbench/README.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+   |AWS Region                |     Link        |
+   |:------------------------:|:-----------:|
+   |us-east-1 (N. Virginia)    | [<img src="./img/CFT.png">](https://console.aws.amazon.com/cloudformation/home?region=us-east-1#/stacks/new?stackName=fmbench&templateURL=https://aws-blogs-artifacts-public.s3.amazonaws.com/artifacts/ML-FMBT/template.yml) |
+   |us-west-2 (Oregon)    | [<img src="./img/CFT.png">](https://console.aws.amazon.com/cloudformation/home?region=us-west-2#/stacks/new?stackName=fmbench&templateURL=https://aws-blogs-artifacts-public.s3.amazonaws.com/artifacts/ML-FMBT/template.yml) |
+
+1. Once the CloudFormation stack is created, navigate to SageMaker Notebooks and open the `fmbench-notebook`.
```
[the remark this answers is not in the document]
Comment (aarora79) [reply recipes/benchmarks/fmbench/README.md:54]: Sorry could not follow, markdown takes care of auto-incrementing the numbers, maybe I am misunderstanding the comment could you please clarify. Thanks.

File: recipes/benchmarks/fmbench/README.md (hunk 2 above)
Comment (aarora79) [reply recipes/benchmarks/fmbench/README.md:64]: image
Sorry could not follow, the attached image is what I see.

File: recipes/benchmarks/fmbench/README.md (hunk 3 above)
Comment (aarora79) [reply recipes/benchmarks/fmbench/README.md:68]: > can we clarify this part a bit, are we packing the requests at 1k tokens? how batching works here? is it packing instead of batching?

No we are not limiting the requests at 1k tokens, what we are saying here is that if you do not provide a custom tokenizer for the model under test then we use the relation 750 words is 1000 tokens to count tokens (token counting is needed to filter prompts if we want to limit prompts within a given size limit, and also for metrics calculation).

Comment (aarora79) [reply recipes/benchmarks/fmbench/README.md:68]: > instructions provided later in this document on how to use a custom

Thank you for pointing this out, will rephrase this to point to the `Tokenizer Directory` section [here](https://github.com/aws-samples/foundation-model-benchmarking-tool/tree/main?tab=readme-ov-file#the-diy-version-with-gory-details).

File: recipes/benchmarks/fmbench/README.md (hunk 4 above)
Comment (aarora79) [reply recipes/benchmarks/fmbench/README.md:84]: ok let me mention that somewhere at the start as well.

File: recipes/benchmarks/fmbench/README.md (hunk 5 above)
Comment (aarora79) [reply recipes/benchmarks/fmbench/README.md:107]: Yes, will add a section for that, the [config files for Llama3 on Bedrock](https://github.com/aws-samples/foundation-model-benchmarking-tool/blob/main/src/fmbench/configs/config-bedrock-llama3.yml) are now available in `FMBench`.

**Review by HamidShojanazeri — APPROVED** [review]
